lang/syntax/src/common/de_bruijn.rs

```diff
@@ -77,24 +77,42 @@ pub trait Leveled {
 }
 
 /// De-Bruijn shifting
-pub trait Shift {
-    /// Shift a term in the first component of the two-dimensional De-Bruijn index
-    fn shift(&self, by: (isize, isize)) -> Self;
-}
-
-pub trait ShiftRange: RangeBounds<usize> + Clone {}
-
-impl<T: RangeBounds<usize> + Clone> ShiftRange for T {}
+///
+/// When we manipulate terms using de Bruijn notation we often
+/// have to change the de Bruijn indices of the variables inside
+/// a term. This is what the "shift" and "shift_in_range" functions
+/// from this trait are for.
+///
+/// Simplified Example: Consider the lambda calculus with de Bruijn
+/// indices whose syntax is "e := n | λ_. e | e e". The shift_in_range
+/// operation would be defined as follows:
+/// - n.shift_in_range(range, by) = if (n ∈ range) then { n + by } else { n }
+/// - (λ_. e).shift_in_range(range, by) = λ_.(e.shift_in_range(range.left += 1, by))
+/// - (e1 e2).shift_in_range(range, by) = (e1.shift_in_range(range, by)) (e2.shift_in_range(range, by))
+/// So whenever we traverse a binding occurrence we have to bump the left
+/// side of the range by one.
+///
+/// Note: We use two-level de Bruijn indices. The cutoff-range only applies to
+/// the first element of a two-level de Bruijn index.
+///
+/// Ref: https://www.cs.cornell.edu/courses/cs4110/2018fa/lectures/lecture15.pdf
+pub trait Shift: Sized {
+    /// Shift all open variables in `self` by the the value indicated with the
+    /// `by` argument.
+    fn shift(&self, by: (isize, isize)) -> Self {
+        self.shift_in_range(0.., by)
+    }
 
-pub trait ShiftInRange {
+    /// Shift every de Bruijn index contained in `self` by the value indicated
+    /// with the `by` argument. De Bruijn indices whose first component lie within
+    /// the indicated `range` are not affected by the shift.
```

Comment:
Only the indices whose first component lie within the indicated `range` *are* affected by the shift.